src/daft-core/src/array/ops/arrow2/sort/primitive/common.rs

```diff
@@ -72,16 +79,33 @@ where
         if descending {
             let mut nulls = 0;
             let mut valids = 0;
+            let last_valid_index = length.saturating_sub(validity.unset_bits());
+
             validity
                 .iter()
                 .zip(I::range(0, length).unwrap())
-                .for_each(|(is_valid, index)| {
-                    if is_valid {
-                        indices[validity.unset_bits() + valids] = index;
-                        valids += 1;
-                    } else {
-                        indices[nulls] = index;
-                        nulls += 1;
+                .for_each(|(is_not_null, index)| {
+                    match (is_not_null, nulls_first) {
+                        // value && nulls first
+                        (true, true) => {
+                            indices[validity.unset_bits() + valids] = index;
+                            valids += 1;
+                        }
```

Comment:
Does `validity.unset_bits()` ever change?

Reply:
no, we could probably set it to a variable instead of inside the loop.